### PR TITLE
Add xege/20.08

### PR DESCRIPTION
**Pull request by Guyutongxue**
Specify library name and version:  **xege/20.08**

`xege`, or Easy Graphics Engine (EGE), is a graphics library in **Windows**. It implements most of the features of [BGI](https://en.wikipedia.org/wiki/Borland_Graphics_Interface). The reason this library was created is that some areas (China, India, etc.) still widely use BGI-like graphics libraries in education, but BGI has been eliminated. This library will be helpful for them.

I'm not the author of this library (the main authors of it are @misakamm and @wysaid). But I'd like to use this library in some projects which I prefer Conan as a package manager. 

*I've tested this package under Windows with MSVC and MinGW-w64.*

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/xege/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    name = "xege"
+    version = "20.08"
+    license = "LGPLv2.1"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://xege.org/"
+    description = "Easy Graphics Engine, a lite graphics library in Windows"
+    topics = ("ege", "graphics", "gui")
+    settings = "os", "compiler", "build_type", "arch"
+    options = {"shared": [True, False]}
+    default_options = {"shared": False}
```
Comment (SpaceIm) [inline recipes/xege/all/conanfile.py:16]: always static: https://github.com/wysaid/xege/blob/40bca13799e512b14570c41f3d285eca616ca9b1/CMakeLists.txt#L5

File: recipes/xege/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        # in MSVC if the packaged project doesn't have variables to set it
+        # properly
+#         tools.replace_in_file("hello/CMakeLists.txt", "PROJECT(HelloWorld)",
+#                               '''PROJECT(HelloWorld)
+# include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+# conan_basic_setup()''')
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure(source_folder="source_subfolder")
```
Comment (SpaceIm) [inline recipes/xege/all/conanfile.py:38]:
```suggestion
        cmake.configure()
```

a CMake wrapper calling `conan_basic_setup()` must be added

File: recipes/xege/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            self.cpp_info.system_libs = [
+                "gdiplus",
+                "uuid",
+                "msimg32",
+                "gdi32",
+                "imm32",
+                "ole32",
+                "oleaut32",
+                "winmm"
+            ]
```
Comment (SpaceIm) [inline recipes/xege/all/conanfile.py:69]: you can add those system libs unconditionally

File: recipes/xege/all/test_package/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+add_executable(example example.cpp)
+target_link_libraries(example ${CONAN_LIBS})
+
+if(MSYS OR MINGW)
+    target_link_libraries(example -static-libgcc -static-libstdc++)
+endif()
```
Comment (SpaceIm) [inline recipes/xege/all/test_package/CMakeLists.txt:12]: why?

Comment (Guyutongxue) [reply recipes/xege/all/test_package/CMakeLists.txt:12]: When I'm testing this under 32bit MinGW-w64, I got some unexpected linkage error to the standard library. So I just use static linkage to prevent them.

This problem doesn't appear in 64bit MinGW, and I don't know why it will happen.

**Comment by Guyutongxue** [comment]
Sorry. It seems I messed it up

File: recipes/xege/all/test_package/CMakeLists.txt (hunk 4 above)
Comment (Guyutongxue) [reply recipes/xege/all/test_package/CMakeLists.txt:12]: Seems that these options are enabled in default for those IDE using 32bit MinGW:
https://github.com/Embarcadero/Dev-Cpp/blob/master/Source/devCFG.pas#L1297

File: recipes/xege/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
+from conans import ConanFile, CMake, tools
+from conans.errors import ConanInvalidConfiguration
+import os
+
+
+class XegeConan(ConanFile):
+    name = "xege"
+    version = "20.08"
```
Comment (jgsogo) [inline recipes/xege/all/conanfile.py:8]:
```suggestion
```

Not needed here, the version will be taken from the `config.yml` file. Without it, this same recipe will serve for other versions as well.

File: recipes/xege/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    license = "LGPLv2.1"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://xege.org/"
+    description = "Easy Graphics Engine, a lite graphics library in Windows"
+    topics = ("ege", "graphics", "gui")
+    settings = "os", "compiler", "build_type", "arch"
+    options = {"shared": [False, True]}
+    default_options = {"shared": False}
```
Comment (jgsogo) [inline recipes/xege/all/conanfile.py:16]:
```suggestion
```

There is no need to enumerate this option if the library is always static.

File: recipes/xege/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    generators = "cmake"
+    exports_sources = ["CMakeLists.txt"]
+
+    def configure(self):
+        if self.options.shared:
+            raise ConanInvalidConfiguration(
+                "This library is always static")
```
Comment (jgsogo) [inline recipes/xege/all/conanfile.py:23]:
```suggestion
```

See comment above.

File: recipes/xege/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = self.name + "-" + self.version
+        os.rename(extracted_dir, "source_subfolder")
+        # This small hack might be useful to guarantee proper /MT /MD linkage
+        # in MSVC if the packaged project doesn't have variables to set it
+        # properly
+#         tools.replace_in_file("hello/CMakeLists.txt", "PROJECT(HelloWorld)",
+#                               '''PROJECT(HelloWorld)
+# include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+# conan_basic_setup()''')
```
Comment (jgsogo) [inline recipes/xege/all/conanfile.py:38]:
```suggestion
```

Not needed

File: recipes/xege/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+        # Explicit way:
+        # self.run('cmake %s/hello %s'
+        #          % (self.source_folder, cmake.command_line))
+        # self.run("cmake --build . %s" % cmake.build_config)
```
Comment (jgsogo) [inline recipes/xege/all/conanfile.py:48]:
```suggestion
```

Not needed

File: recipes/xege/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
+from conans import ConanFile, CMake, tools
+from conans.errors import ConanInvalidConfiguration
+import os
+
+
+class XegeConan(ConanFile):
+    name = "xege"
```
Comment (prince-chrismc) [inline recipes/xege/all/conanfile.py:7]: Seems this is a fork

https://github.com/wysaid/xege

```suggestion
    name = "wysaid-xege"
    provides = "xege"
```

File: recipes/xege/all/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
+from conans import ConanFile, CMake, tools
+from conans.errors import ConanInvalidConfiguration
+import os
+
```
Comment (prince-chrismc) [inline recipes/xege/all/conanfile.py:4]: For strip_root

```suggestion

required_conan_version = ">=1.33.0"

```

File: recipes/xege/all/conanfile.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    def configure(self):
+        if self.settings.os != "Windows":
+            raise ConanInvalidConfiguration(
+                "This library is only compatible for Windows")
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = self.name + "-" + self.version
+        os.rename(extracted_dir, "source_subfolder")
```
Comment (prince-chrismc) [inline recipes/xege/all/conanfile.py:25]:
```suggestion
        tools.get(**self.conan_data["sources"][self.version], destination=self._source_subfolder, strip_root=True)
```

File: recipes/xege/all/conanfile.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+            self.cpp_info.libs = ["graphics"]
+        self.cpp_info.system_libs = [
+            "gdiplus",
+            "uuid",
+            "msimg32",
+            "gdi32",
+            "imm32",
+            "ole32",
+            "oleaut32",
+            "winmm"
```
Comment (prince-chrismc) [inline recipes/xege/all/conanfile.py:53]: I might have miscounted but there's one extra? https://github.com/wysaid/xege/blob/master/src/CMakeLists.txt#L53

```suggestion
            "gdiplus",
            "uuid",
            "msimg32",
            "gdi32",
            "imm32",
            "ole32",
            "oleaut32"
```

File: recipes/xege/all/test_package/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+
+# CTest is a testing tool that can be used to test your project.
+# enable_testing()
+# add_test(NAME example
+#          WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/bin
+#          COMMAND example)
```
Comment (prince-chrismc) [inline recipes/xege/all/test_package/CMakeLists.txt:18]:
```suggestion
```

File: recipes/xege/all/test_package/conanfile.py [hunk 15]
```diff
@@ -0,0 +1,19 @@
+# Tested under MinGW-w64 32bit / 64bit and Visual Studio 2019.
+
+from conans import ConanFile, CMake, tools
```
Comment (prince-chrismc) [inline recipes/xege/all/test_package/conanfile.py:3]:
```suggestion
from conans import ConanFile, CMake, tools
```

File: recipes/xege/all/test_package/example.cpp [hunk 16]
```diff
@@ -0,0 +1,15 @@
+#include <ege.h>
+#include <windows.h>
+
+#ifdef _MSC_VER
+#pragma comment(lib,"msvcrtd.lib")
```
Comment (prince-chrismc) [inline recipes/xege/all/test_package/example.cpp:5]: WinDDK debug runtimes 😕 Why is this required?

File: recipes/xege/all/conanfile.py (hunk 10 above)
Comment (prince-chrismc) [reply recipes/xege/all/conanfile.py:7]: Seems weird the "upstream" has no code

File: recipes/xege/all/test_package/example.cpp (hunk 16 above)
Comment (Guyutongxue) [reply recipes/xege/all/test_package/example.cpp:5]: I don't know too. But if I don't do this, Conan CI will throw out a likage error comppains about `__imp_*`ref not found.

This problem doesn't appear during my local test.

File: recipes/xege/all/conanfile.py (hunk 10 above)
Comment (Guyutongxue) [reply recipes/xege/all/conanfile.py:7]: Yes, the original repo of this library doesn't update anymore. And `wysaid` improved it a lot, like building with CMake, providing offical website, etc.

So It's necessary to change the package name?

Comment (prince-chrismc) [reply recipes/xege/all/conanfile.py:9]: No. Well see what others think but I fine for me

File: recipes/xege/all/test_package/example.cpp (hunk 16 above)
Comment (prince-chrismc) [reply recipes/xege/all/test_package/example.cpp:5]: This code was in the very first commit... can you point me to the error you saw?

Comment (Guyutongxue) [reply recipes/xege/all/test_package/example.cpp:5]: Sorry, I've started a pr #5972 at an earlier time, and [here](https://github.com/conan-io/conan-center-index/pull/5972#issuecomment-864108888) I found this error. After that, I closed it and change a new example.

Comment (prince-chrismc) [reply recipes/xege/all/test_package/example.cpp:5]: Ah since you missed the cmake wrapper and it was only MT that fails... its possible this is no longer required (or maybe add it to the system libs so it works for consumers)

Comment (Guyutongxue) [reply recipes/xege/all/test_package/example.cpp:5]: I will check it now. Thanks for your advice.

**Review by prince-chrismc — APPROVED** [review]

File: recipes/xege/all/conanfile.py (hunk 10 above)
Comment (jgsogo) [reply recipes/xege/all/conanfile.py:9]: I think it is fine too, the fork really looks like the official EGE

**Review by jgsogo — APPROVED** [review]

File: recipes/xege/all/test_package/CMakeLists.txt (hunk 4 above)
Comment (SpaceIm) [reply recipes/xege/all/test_package/CMakeLists.txt:12]: it's odd, but isolated in test package.

**Review by SpaceIm — APPROVED** [review]
